datatable/java/datatable/src/main/java/io/cucumber/datatable/DataTableTypeRegistry.java

```diff
@@ -26,69 +26,79 @@
+            defineDataTableType(new DataTableType(short.class, shortTableCellTransformer));
+            Type optionalShort = new TypeReference<Optional<Short>>() {}.getType();
+            defineDataTableType(new DataTableType(optionalShort, optionalIfAbsent(shortTableCellTransformer)));
+        }
+        {
+            TableCellTransformer<Integer> integerTableCellTransformer = applyIfPresent(Integer::decode);
+            defineDataTableType(new DataTableType(Integer.class, integerTableCellTransformer));
+            defineDataTableType(new DataTableType(int.class, integerTableCellTransformer));
+            Type optionalInteger = new TypeReference<Optional<Integer>>() {}.getType();
+            defineDataTableType(new DataTableType(optionalInteger, optionalIfAbsent(integerTableCellTransformer)));
+        }
+        {
+            TableCellTransformer<Long> longTableCellTransformer = applyIfPresent(Long::decode);
+            defineDataTableType(new DataTableType(Long.class, longTableCellTransformer));
+            defineDataTableType(new DataTableType(long.class, longTableCellTransformer));
+            Type optionalLong = new TypeReference<Optional<Long>>() {}.getType();
+            defineDataTableType(new DataTableType(optionalLong, optionalIfAbsent(longTableCellTransformer)));
+        }
+        {
+            TableCellTransformer<Float> floatTableCellTransformer = applyIfPresent(numberParser::parseFloat);
+            defineDataTableType(new DataTableType(Float.class, floatTableCellTransformer));
+            defineDataTableType(new DataTableType(float.class, floatTableCellTransformer));
+            Type optionalFloat = new TypeReference<Optional<Float>>() {}.getType();
+            defineDataTableType(new DataTableType(optionalFloat, optionalIfAbsent(floatTableCellTransformer)));
+        }
+        {
+            TableCellTransformer<Double> doubleTableCellTransformer = applyIfPresent(numberParser::parseDouble);
+            defineDataTableType(new DataTableType(Double.class, doubleTableCellTransformer));
+            defineDataTableType(new DataTableType(double.class, doubleTableCellTransformer));
+            Type optionalDouble = new TypeReference<Optional<Double>>() {}.getType();
+            defineDataTableType(new DataTableType(optionalDouble, optionalIfAbsent(doubleTableCellTransformer)));
+        }
     }
 
     private static <R> TableCellTransformer<R> applyIfPresent(Function<String, R> f) {
         return s -> s == null ? null : f.apply(s);
     }
 
-    private static <T> TableCellTransformer<Optional<T>> transformAsOptional(TableCellTransformer<T> tableCellTransformer) {
-        return s -> s == null || s.isEmpty() ? Optional.empty() : Optional.of(tableCellTransformer.transform(s));
+    private static <R> TableCellTransformer<Optional<R>> optionalIfAbsent(TableCellTransformer<R> t) {
+        return s -> s == null || s.isEmpty() ? Optional.empty() : Optional.ofNullable(t.transform(s));
```

Comment:
While it would be unexpected the function returned by `applyIfPresent` will return null when `f.appy` returns `null`. So `ofNullable` should be used here.